### PR TITLE
Vs30 refactor

### PR DESCRIPTION
**Pull request by AndrewRidden-Harper**
## Re-opened after history rewrite

This PR is a re-opened version of an earlier one. Between then and now, I used `git filter-repo` to remove some unused benchmark files from the repo's history. The rewrite worked as intended but changed every commit's SHA, which orphaned the previous branch from `master` and made the old PR unmergeable, so I closed it and re-targeted the same work against the rewritten history. The content is unchanged; only the underlying SHAs have shifted.
 
> [!IMPORTANT]
> **If you have an existing local clone of this repository, delete it and re-clone from scratch before pulling these changes.** Your local commits share no history with  `origin` anymore - `git pull` on top of an old clone will fail or produce a broken state. A fresh `git clone` is the safe move.
  
## Summary
  
This PR continues the Vs30 package refactor, addressing prior review feedback, adding two new bundled models and support for custom YAML configurations, and incorporating other improvements.

### Addressed review feedback

Several review comments from the previous PR have all been addressed. Most do not warrant a specific call-out but some structural changes are worth flagging:
  
- **Removed CLI access to individual pipeline steps** (interested users can still call the functions)
- **CLI module (`cli.py`) is now just a CLI wrapper** (without pipeline orchestration code)
- **Shared parameters lifted from per-model configs into `constants.py`.** Values that were previously declared in each of the bundled model config files are now defined once in code; the config files are correspondingly leaner.
- **Bundled model versions are specified by a `StrEnum` rather than their config file name.** The four valid names live as `FixedModelVersion` members in `vs30/constants.py`, giving a single canonical source for the supported set.
  
### Additional development work
  
- **Ensured exact correspondence between points and grid modes.** New tests verify identical outputs at tricky locations like geological category boundaries, missing-value edges in the terrain raster, and similar edge cases.
- **Tightened grid-bounds semantics** to ensure unambiguous pixel-edge interpretation and eliminate the potential for inconsistencies from GDAL tie-breaking behaviour.
- **Added the `jaehwi_v1p0` model** (in addition to the existing `modified_foster_2019` and `viktor_cpt_clustering`), along with the gap-filling functionality it requires.
- **Added a near-reproduction of the Foster et al. (2019) model** (`foster_2019_approx`). Exact reproduction wasn't pursued because the published model has known flaws.
- **Added support for custom YAML model configurations.** Users can pass a YAML config path in place of a bundled model name to define their own model variants without modifying the package.
- **Added small end-to-end benchmark tests** with reference outputs generated by the legacy code, to verify reproduction.
- **Removed the Python multiprocessing option.** Testing across a wide range of scenarios showed it was consistently slower than the single-process path, which is already multi-threaded via the BLAS used by NumPy/SciPy.
- **Other performance touch-ups** — einsum-based MVN distances, gapfill memory optimisation, Float32 coast-distance arrays, etc.
- **Added a brief GitHub wiki.**
  
### Why points and grid are separate pipelines

A natural design question is why the package maintains two separate pipelines rather than collapsing them into one. Both pipelines produce identical results at every location (verified by the correspondence tests above); they are kept separate purely for performance. Each direction of unification has its own performance problem:  
- **Have `vs30 grid` call the points pipeline at each pixel.** This loses the bulk array operations that the grid pipeline gets from GDAL: bulk raster reads, bulk sampling, bulk spatial adjustment. At national-grid sizes (millions of pixels) it would be impractically slow.
- **Have `vs30 points` run the grid pipeline and sample the desired sites from it.** This would compute Vs30 across an entire grid regardless of how many sites are requested, making typical points-mode queries needlessly slow.

### CODE REVIEW

File: vs30/category.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+            f"No usable prior columns. Need either "
+            f"{constants.COL_POSTERIOR_MEAN_CLUSTERED}+{constants.COL_POSTERIOR_STDV_CLUSTERED} "
+            f"or {constants.COL_MEAN}+{constants.COL_STDV}."
+        )
+
+    posterior_df[constants.COL_PRIOR_STDV] = np.clip(
+        posterior_df[constants.COL_PRIOR_STDV].to_numpy(),
+        constants.MIN_SIGMA,
+        None,
+    )
```
[the remark this answers is not in the document]
Comment (AndrewRidden-Harper) [reply vs30/category.py:166]: The factual claim — that clipping the prior doesn't guarantee `posterior ≥ MIN_SIGMA` — is correct. The inferred conclusion that the *code* is wrong isn't: the refactor is reproducing legacy behaviour. The legacy `posterior()` does exactly `stdv = np.maximum(model[:, 1], min_sigma)`, then never re-clips. Changing the clip to operate on the posterior would diverge from the legacy output the refactor is pinned against.

So the real defect is documentation, not logic: the comment on `MIN_SIGMA` in `constants.py` claimed "after Bayesian update" when the behaviour is "before". Fixed the comment to describe the actual prior-only clip and note that posteriors can drop below `MIN_SIGMA` when observations strongly agree with the prior mean.

File: vs30/correlations.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        Correlation length parameter in meters.
+    min_dist : float, optional
+        Minimum distance enforced to prevent division issues. Default is MIN_DIST_ENFORCED.
+
+    Returns
+    -------
+    ndarray
+        Correlation values between 0 and 1. Same shape as distances.
+    """
+    return np.exp(-np.maximum(min_dist, distances) / phi)
```
[the remark this answers is not in the document]
Comment (AndrewRidden-Harper) [reply vs30/correlations.py:34]: Declining.

The mathematical observation is correct: with `min_dist = 0.1 m`, `corr(0) = exp(-0.1 / phi) < 1`. The magnitude isn't material though — at `phi = 1407 m` that's `corr(0) ≈ 0.999929`, i.e. ~0.007 % variance shrinkage on the diagonal. The code at `spatial.py:642–644` already acknowledges this explicitly ("≈1 minus a tiny epsilon from the enforced minimum distance").

The suggested replacement (`return np.exp(-distances / phi)`) also has a substantive issue: it removes the safeguard that `MIN_DIST_ENFORCED` exists for in the first place. Two co-located observations would yield `corr = 1.0` exactly, producing a rank-deficient covariance matrix and breaking the MVN solve. The Matérn variant in the same file relies on the same clamp because Bessel `kv` diverges at zero, so an unconditional removal would also break that path.

A "correct" fix (special-casing the corr-matrix diagonal to exactly 1 while keeping the off-diagonal clamp) would address the 0.007 % cosmetic concern but the cost-to-benefit doesn't justify the churn or the additional code path.

File: vs30/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+                "do_bayesian_update": False,
+                "include_intermediate": False,
+                "fill_gaps": False,
+            }
+
+            for idx in fillable_indices:
+                easting, northing = locations[idx]
+                fill_vs30, fill_stdv = fill_one_point_via_local_grid(
+                    easting, northing, gapfill_grid_config, grid_pipeline_kwargs
+                )
```
[the remark this answers is not in the document]
Comment (AndrewRidden-Harper) [reply vs30/pipeline.py:1162]: Declining for now.

Agreed that running a full `grid_pipeline` per fillable point is wasteful at scale. But neither suggested alternative survives contact with the actual donor-lookup flow:
  - "Identify the nearest valid lattice point for all gap points and evaluate in a single batch" presupposes a precomputed valid lattice across the relevant region — which is exactly the work the local-grid approach exists to *avoid*. Producing that lattice means running the full NZ grid pipeline.
- "Use a spatial index to find donor values from the already computed valid points" is chicken-and-egg in points mode: there *are* no already-computed valid points outside the user-supplied locations until the local pipeline runs.

A genuine improvement here is to cluster fillable points that share a local-grid footprint and amortise one pipeline call across the cluster. That's a non-trivial refactor and the savings only kick in when users feed many clustered fillable points, which isn't an established pain point. Happy to revisit if/when someone hits this in practice.

File: vs30/spatial.py [hunk 4]
```diff
@@ -0,0 +1,1055 @@
...
+    obs_data: ObservationData,
+    max_dist_m: float = constants.MAX_DIST_M,
+    max_points: int = constants.MAX_POINTS,
+) -> np.ndarray:
+    """
+    Select observations for a pixel using distance filtering.
+
+    Parameters
+    ----------
+    pixel : PixelData
```
[the remark this answers is not in the document]
Comment (AndrewRidden-Harper) [reply vs30/spatial.py]: Thanks, this is a good suggestion. It has been implemented,

File: vs30/constants.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+STANDARD_ID_COLUMN: str = "id"
+
+# Coordinate Reference System for New Zealand Transverse Mercator 2000
+NZTM_CRS: str = "EPSG:2193"
+
+# Default encoding for reading text files (e.g. YAML configs).
+DEFAULT_TEXT_ENCODING: str = "utf-8"
+
+
+class ObservationColumn:
```
Comment (lispandfound) [inline vs30/constants.py]: Could, perhaps should be an enum?

File: vs30/raster.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+            row_offset = round((g_ymax - s_ymax) / dy)
+            distance_meters = data[
+                row_offset : row_offset + template_profile["height"],
+                col_offset : col_offset + template_profile["width"],
+            ].astype(np.float32)
+        else:
+            distance_meters = data.astype(np.float32)
+    finally:
+        Path(tmp_path).unlink(missing_ok=True)
+
```
Comment (lispandfound) [inline vs30/raster.py]: You should use a with statement plus `tempfile.NamedTemporaryFile`.

File: vs30/resources/observations/README.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+- **Source label**: `mcgann`
+
+#### 2. Wotherspoon et al. (2015)
+- **Source**: `sites_load.load_wotherspoon_vs()`
+- **Description**: Additional surface wave measurements from Canterbury
+- **Count**: ~40 sites
+- **Source label**: `wotherspoon`
+
+#### 3. Kaiser et al. (2017) - Filtered
+- **Source**: `sites_load.load_kaiseretal_vs()`
```
Comment (joelridden) [inline vs30/resources/observations/README.md]: sites_load.py doesn't exist anymore? Not sure where the functions live now

Comment (AndrewRidden-Harper) [reply vs30/resources/observations/README.md]: Good catch — this README hadn't had a careful pass before the PR, and following up on your comment surfaced enough staleness (and a few inaccuracies) that I rewrote the whole thing (ea74b28). `sites_load.py` was legacy and removed in the refactor, so those loaders aren't in the package anymore; the CSVs now ship pre-generated, with `sites_load.py` kept only as a provenance note for where the data came from.
